core/src/main/java/com/github/phantazmnetwork/core/player/BasicPlayerView.java

```diff
@@ -123,6 +127,35 @@ private CompletableFuture<String> getUsernameRequest() {
         return Optional.ofNullable(player);
     }
 
+    @Override
+    public @NotNull Optional<PlayerInfo> getPlayerInfo() {
+        Optional<Player> playerOptional = getPlayer();
+        if (playerOptional.isEmpty()) {
+            invalidatePlayerInfo();
+            return Optional.empty();
+        }
+
+        if (playerInfo != null) {
+            return Optional.of(playerInfo);
+        }
+
+        if (protocolVersion == -1) {
+            return Optional.empty();
+        }
+
+        playerInfo = new PlayerInfo(playerOptional.get(), protocolVersion);
+        return Optional.of(playerInfo);
+    }
+
+    public void invalidatePlayerInfo() {
+        playerInfo = null;
+        protocolVersion = -1;
+    }
+
+    public void setProtocolVersion(int protocolVersion) {
+        this.protocolVersion = protocolVersion;
```

Comment:
IllegalStateException?

Reply:
i'm mostly just wondering why it needs to reset in the first place

Reply:
Player goes online 1.18
Player protocol syncs 1.18
Player goes offline
Player goes online 1.8
Code requests protocol version, gets 1.18
Player protocol syncs 1.8

it's for that small timeframe where plugin messages aren't sent

Reply:
from the POV of the server, the player is never *actually* on 1.8 though?

Reply:
we might need it to stop 1.8 players from joining a map with the spyglass gun

Reply:
it determines the actual version only for the purposes of us making a comparison, but it doesn't actually need to be reset for offline players

Reply:
isn't it just going to be set to the (correct) value when they rejoin

Reply:
in the order of events I sent it's outdated

Reply:
a) have you actually observed this "small timeframe where plugin messages aren't sent?" would the player even be considered online at that point?
b) if so, surely there's a better way to send the data

Reply:
the timeframe starts from the beginning of the PLAY phase (so they are registered) until the plugin message packet is sent
we could stuff this into the handshake packet, but really this is what the plugin message packet is designed for

Reply:
> really this is what the plugin message packet is designed for
i would agree if we were forwarding pretty much anything *except* player protocol version

also it seems slightly hard to believe that via* doesn't offer some way to forward protocol version

Reply:
I've logged all plugin messages previously, nothing from via + I didn't see a protocol version in the handshake when I was testing bungee auth + I don't see any config option to turn that on

Reply:
just checked the source and can confirm that there appears to be no hidden forwarding API

Reply:
i still think this needs to go in the handshake packet though

Reply:
it seems possible to modify the player gameprofile properties before the handshake is sent on velocity's side
we would need to create our own custom event on minestom's side to handle the handshake packet since there is no such event and minestom has no concept of gameprofile properties

I don't particularly like adding it to the handshake though, it seems unnecessary

Reply:
minecraft also sends its brand as a plugin message, which is in a similar vein of information / use case as the protocol version
therefore I think it'd be best to keep the protocol version as a plugin message too

Reply:
brands are basically cosmetic-only though, they don't actually matter